File: mapreduce/yt/skiff/skiff_schema.cpp
#include "skiff_schema.h"
#include <util/digest/numeric.h>
#include <util/str_stl.h>

////////////////////////////////////////////////////////////////////////////////

size_t THash<NSkiff::TSkiffSchema>::operator()(const NSkiff::TSkiffSchema &schema) const
{
    auto hash = CombineHashes(
        THash<TString>()(schema.GetName()),
        static_cast<size_t>(schema.GetWireType()));
    for (const auto& child : schema.GetChildren()) {
        hash = CombineHashes(hash, (*this)(*child));
    }
    return hash;
}

////////////////////////////////////////////////////////////////////////////////

namespace NSkiff {

////////////////////////////////////////////////////////////////////////////////

bool operator==(const TSkiffSchema& lhs, const TSkiffSchema& rhs)
{
    if (lhs.GetChildren().size() != rhs.GetChildren().size()
        || lhs.GetName() != rhs.GetName()
        || lhs.GetWireType() != rhs.GetWireType())
    {
        return false;
    }
    return std::equal(
        lhs.GetChildren().begin(),
        lhs.GetChildren().end(),
        rhs.GetChildren().begin(),
        TSkiffSchemaPtrEqual());
}

////////////////////////////////////////////////////////////////////////////////

void PrintShortDebugString(const TSkiffSchema& schema, IOutputStream* out)
{
    (*out) << ToString(schema.GetWireType());
    if (!IsSimpleType(schema.GetWireType())) {
        auto children = schema.GetChildren();
        if (!children.empty()) {
            (*out) << '<';
            for (const auto& child : children) {
                PrintShortDebugString(*child, out);
                (*out) << ';';
            }
            (*out) << '>';
        }
    }
}

void PrintShortDebugString(const TSkiffSchemaPtr& schema, IOutputStream* out)
{
    PrintShortDebugString(*schema, out);
}

TString GetShortDebugString(const TSkiffSchema& schema)
{
    TStringStream out;
    PrintShortDebugString(schema, &out);
    return out.Str();
}

TString GetShortDebugString(const TSkiffSchemaPtr& schema)
{
    TStringStream out;
    PrintShortDebugString(schema, &out);
    return out.Str();
}

TSimpleTypeSchemaPtr CreateSimpleTypeSchema(EWireType type)
{
    return ::MakeIntrusive<TSimpleTypeSchema>(type);
}

static void VerifyNonemptyChildren(const TSkiffSchemaList& children, EWireType wireType)
{
    if (children.empty()) {
        ythrow yexception() << wireType << " must have at least one child";
    }
}

TTupleSchemaPtr CreateTupleSchema(TSkiffSchemaList children)
{
    VerifyNonemptyChildren(children, EWireType::Tuple);
    return ::MakeIntrusive<TTupleSchema>(std::move(children));
}

TVariant8SchemaPtr CreateVariant8Schema(TSkiffSchemaList children)
{
    VerifyNonemptyChildren(children, EWireType::Variant8);
    return ::MakeIntrusive<TVariant8Schema>(std::move(children));
}

TVariant16SchemaPtr CreateVariant16Schema(TSkiffSchemaList children)
{
    VerifyNonemptyChildren(children, EWireType::Variant16);
    return ::MakeIntrusive<TVariant16Schema>(std::move(children));
}

TRepeatedVariant8SchemaPtr CreateRepeatedVariant8Schema(TSkiffSchemaList children)
{
    VerifyNonemptyChildren(children, EWireType::RepeatedVariant8);
    return ::MakeIntrusive<TRepeatedVariant8Schema>(std::move(children));
}

TRepeatedVariant16SchemaPtr CreateRepeatedVariant16Schema(TSkiffSchemaList children)
{
    VerifyNonemptyChildren(children, EWireType::RepeatedVariant16);
    return ::MakeIntrusive<TRepeatedVariant16Schema>(std::move(children));
}

////////////////////////////////////////////////////////////////////////////////

TSkiffSchema::TSkiffSchema(EWireType type)
    : Type_(type)
{ }

EWireType TSkiffSchema::GetWireType() const
{
    return Type_;
}

TSkiffSchemaPtr TSkiffSchema::SetName(TString name)
{
    Name_ = std::move(name);
    return this;
}

const TString& TSkiffSchema::GetName() const
{
    return Name_;
}

const TSkiffSchemaList& TSkiffSchema::GetChildren() const
{
    static const TSkiffSchemaList schema = {};
    return schema;
}

////////////////////////////////////////////////////////////////////////////////

TSimpleTypeSchema::TSimpleTypeSchema(EWireType type)
    : TSkiffSchema(type)
{
    Y_ENSURE(IsSimpleType(type));
}

////////////////////////////////////////////////////////////////////////////////

} // namespace NSkiff

File: yt/java/skiff/src/main/java/tech/ytsaurus/skiff/SkiffSchema.java
package tech.ytsaurus.skiff;

import java.util.List;
import java.util.Objects;
import java.util.Optional;

import javax.annotation.Nullable;

import tech.ytsaurus.ysontree.YTreeBuilder;
import tech.ytsaurus.ysontree.YTreeNode;

import static tech.ytsaurus.ysontree.YTree.listBuilder;
import static tech.ytsaurus.ysontree.YTree.mapBuilder;

public abstract class SkiffSchema {
    protected final WireType type;
    @Nullable
    private String name;

    SkiffSchema(WireType type) {
        this.type = type;
    }

    public WireType getWireType() {
        return type;
    }

    public SkiffSchema setName(String name) {
        this.name = name;
        return this;
    }

    public Optional<String> getName() {
        return Optional.ofNullable(name);
    }

    public abstract List<SkiffSchema> getChildren();

    public abstract boolean isListSchema();

    public abstract boolean isMapSchema();

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        SkiffSchema that = (SkiffSchema) o;
        return Objects.equals(type, that.getWireType()) &&
                Objects.equals(name, that.name) &&
                Objects.equals(getChildren(), that.getChildren());
    }

    @Override
    public int hashCode() {
        return Objects.hash(type, name, getChildren());
    }

    public YTreeNode toYTree() {
        YTreeBuilder ytreeBuilder = mapBuilder()
                .key("wire_type").value(type.toString());
        if (name != null) {
            ytreeBuilder.key("name").value(name);
        }
        if (!getChildren().isEmpty()) {
            YTreeBuilder listBuilder = listBuilder();
            for (SkiffSchema child : getChildren()) {
                listBuilder.value(child.toYTree());
            }
            ytreeBuilder.key("children").value(listBuilder.endList().build());
        }
        return ytreeBuilder.endMap().build();
    }

    public static SimpleTypeSchema simpleType(WireType type) {
        return new SimpleTypeSchema(type);
    }

    public static ComplexSchema tuple(List<SkiffSchema> children) {
        return new ComplexSchema(WireType.TUPLE, children);
    }

    public static ComplexSchema variant8(List<SkiffSchema> children) {
        verifyNonEmptyChildren(children, WireType.VARIANT_8);
        return new ComplexSchema(WireType.VARIANT_8, children);
    }

    public static ComplexSchema variant16(List<SkiffSchema> children) {
        verifyNonEmptyChildren(children, WireType.VARIANT_16);
        return new ComplexSchema(WireType.VARIANT_16, children);
    }

    public static ComplexSchema repeatedVariant8(List<SkiffSchema> children) {
        verifyNonEmptyChildren(children, WireType.REPEATED_VARIANT_8);
        return new ComplexSchema(WireType.REPEATED_VARIANT_8, children);
    }

    public static ComplexSchema repeatedVariant16(List<SkiffSchema> children) {
        verifyNonEmptyChildren(children, WireType.REPEATED_VARIANT_16);
        return new ComplexSchema(WireType.REPEATED_VARIANT_16, children);
    }

    public static SimpleTypeSchema nothing() {
        return new SimpleTypeSchema(WireType.NOTHING);
    }

    private static void verifyNonEmptyChildren(List<SkiffSchema> children, WireType type) {
        if (children.isEmpty()) {
            throw new IllegalArgumentException("\"" + type + "\"must have at least one child");
        }
    }
}
